File: solo-leveling-player-service/src/main/java/com/sleepkqq/sololeveling/player/service/mapper/EnumOrdinalMapper.java
package com.sleepkqq.sololeveling.player.service.mapper;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.JsonDeserializer;
import com.fasterxml.jackson.databind.JsonSerializer;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializerProvider;
import com.fasterxml.jackson.databind.module.SimpleModule;
import java.io.IOException;

public class EnumOrdinalMapper {

  public static <E extends Enum<E>> ObjectMapper createMapper(Class<E> enumType) {
    var mapper = new ObjectMapper();
    var module = new SimpleModule();

    module.addSerializer(enumType, new JsonSerializer<Enum<?>>() {
      @Override
      public void serialize(Enum<?> value, JsonGenerator gen, SerializerProvider provider)
          throws IOException {
        gen.writeNumber(value.ordinal());
      }
    });

    module.addDeserializer(enumType, new JsonDeserializer<>() {
      @Override
      public E deserialize(JsonParser p, DeserializationContext ctx)
          throws IOException {
        var ordinal = p.getIntValue();
        return enumType.getEnumConstants()[ordinal];
      }
    });

    mapper.registerModule(module);
    return mapper;
  }
}

File: solo-leveling-player-service/src/main/kotlin/com/sleepkqq/sololeveling/player/service/mapper/EnumOrdinalMapper.kt
package com.sleepkqq.sololeveling.player.service.mapper

import com.fasterxml.jackson.core.JsonGenerator
import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.databind.*
import com.fasterxml.jackson.databind.module.SimpleModule

object EnumOrdinalMapper {

	fun <E : Enum<E>> createMapper(enumType: Class<E>): ObjectMapper {
		val mapper = ObjectMapper()
		val module = SimpleModule()

		module.addSerializer(
			enumType,
			object : JsonSerializer<Enum<E>>() {
				override fun serialize(
					value: Enum<E>,
					gen: JsonGenerator,
					provider: SerializerProvider
				) = gen.writeNumber(value.ordinal)
			}
		)

		module.addDeserializer(
			enumType,
			object : JsonDeserializer<E>() {
				override fun deserialize(p: JsonParser, ctx: DeserializationContext): E =
					enumType.enumConstants[p.intValue]
			}
		)

		mapper.registerModule(module)
		return mapper
	}
}
